### PR TITLE
MT Gauss-Seidel fixups

### PR DESCRIPTION
**Pull request by cgcgcg**
- Add damping option
- Instead of inverting the diagonal in every iteration, save the inverse of the diagonal
- Allow to pass a precomputed diagonal. 

The last item is necessary for MT GS to work correctly with Ifpack2. The diagonal of a Tpetra matrix might not be given by pairs (i, i) of local IDs (depending on the rowmap and the colmap). Moreover, this allows to use MT GS with the l1 option from Ifpack2.

I also added some routines to dump to MatrixMarket format.

This addresses Issues #221 and #240

### CODE REVIEW

File: src/common/KokkosKernels_SparseUtils.hpp [hunk 1]
```diff
@@ -2133,7 +2133,117 @@ void kk_reduce_numrows_larger_than_threshold(size_t num_elements, view_type view
...
+
+  // Print the Matrix Market banner line.  We assume
+  // nonsymmetric storage ("general").
+
+  out << "%%MatrixMarket matrix array " << dataType << " general" << endl;
+
+  // Write the dimensions of the sparse matrix: (# rows, #
+  // columns, # matrix entries (counting duplicates as
+  // separate entries)).
+  out << numRows << " " << 1 << " " << endl;
```
Comment (srajama1) [inline src/common/KokkosKernels_SparseUtils.hpp]: Sorry to nit pick, space in the end of line ?

Comment (srajama1) [inline src/common/KokkosKernels_SparseUtils.hpp]: Sorry to nit pick, space in the end of line ?

File: src/sparse/KokkosSparse_gauss_seidel_handle.hpp [hunk 2]
```diff
@@ -247,12 +247,12 @@ class GaussSeidelHandle{
   void set_old_to_new_map(const nnz_lno_persistent_work_view_t &old_to_new_map_) {
     this->old_to_new_map = old_to_new_map_;
   }
-  void set_permuted_diagonals (const scalar_persistent_work_view_t permuted_diagonals_){
-    this->permuted_diagonals = permuted_diagonals_;
+  void set_permuted_inverse_diagonal (const scalar_persistent_work_view_t permuted_inverse_diagonal_){
```
Comment (srajama1) [inline src/sparse/KokkosSparse_gauss_seidel_handle.hpp]: Does Ifpack2 already call set_permuted_diagonals ?

File: src/sparse/KokkosSparse_gauss_seidel.hpp [hunk 3]
```diff
@@ -193,7 +193,75 @@ namespace Experimental{
...
+			  "KokkosSparse::gauss_seidel_symbolic: Size type of the matrix should be same as kernelHandle sizetype.");
+
+	  static_assert (std::is_same<typename KernelHandle::const_nnz_lno_t,
+			  typename lno_nnz_view_t_::const_value_type>::value,
+			  "KokkosSparse::gauss_seidel_symbolic: lno type of the matrix should be same as kernelHandle lno_t.");
+
+	  static_assert (std::is_same<typename KernelHandle::const_nnz_scalar_t,
+	  			  typename scalar_nnz_view_t_::const_value_type>::value,
+	  			  "KokkosSparse::gauss_seidel_symbolic: scalar type of the matrix should be same as kernelHandle scalar_t.");
+
```
Comment (srajama1) [inline src/sparse/KokkosSparse_gauss_seidel.hpp]: This is mainly pattern match, so I assume this is ok.

File: src/sparse/KokkosSparse_gauss_seidel_handle.hpp (hunk 2 above)
Comment (cgcgcg) [reply src/sparse/KokkosSparse_gauss_seidel_handle.hpp]: No, it doesn't. This method only sets the permuted diagonal, it does not do any permutations.

File: src/common/KokkosKernels_SparseUtils.hpp [hunk 4]
```diff
@@ -2133,7 +2133,117 @@ void kk_reduce_numrows_larger_than_threshold(size_t num_elements, view_type view
...
+    if (KAM::isComplex) {
+      out << KAM::real (val) << " " << KAM::imag (val);
+    } else {
+      out << val;
+    }
+    out << endl;
+    }
+  }
+
+
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply src/common/KokkosKernels_SparseUtils.hpp]: Oh, I didn't see those! Then there is obviously no reason to add this again.

**Comment by cgcgcg** [comment]
@srajama1 I removed the matrix market routines and took care of all the whitespace and indentation issues. This blows up the diff quite significantly, but there are no functional changes. I ran the unit test on white and bowman. Anything else I should be running?

File: src/sparse/KokkosSparse_gauss_seidel.hpp [hunk 5]
```diff
@@ -49,611 +49,689 @@
...
+      Internal_alno_row_view_t_ const_a_r (row_map.data(), row_map.extent(0));
+      Internal_alno_nnz_view_t_ const_a_l (entries.data(), entries.extent(0));
+      Internal_ascalar_nnz_view_t_ const_a_v (values.data(), values.extent(0));
+      Internal_ascalar_nnz_view_t_ const_a_d (given_inverse_diagonal.data(), given_inverse_diagonal.extent(0));
+
+
+      using namespace KokkosSparse::Impl;
+
+      GAUSS_SEIDEL_NUMERIC<const_handle_type, Internal_alno_row_view_t_, Internal_alno_nnz_view_t_, Internal_ascalar_nnz_view_t_>::gauss_seidel_numeric
+        (&tmp_handle, num_rows, num_cols, const_a_r, const_a_l, const_a_v, const_a_d, is_graph_symmetric);
```
Comment (cgcgcg) [inline src/sparse/KokkosSparse_gauss_seidel.hpp]: @srajama1 The diagonal is passed here as `const_a_d`.

Comment (srajama1) [reply src/sparse/KokkosSparse_gauss_seidel.hpp]: Thanks, missed the basic thing.

**Comment by srajama1** [comment]
Will review this after a meeting.

**Comment by srajama1** [comment]
A unit test in KK will be nice as it gets tested on lot more compilers and platforms than Ifpack2.

**Comment by cgcgcg** [comment]
@srajama1 I changed the unit test so it would run with damping. I'll build and test on Bowman and White..

**Comment by cgcgcg** [comment]
@srajama1 Tests pass on Bowman and White.

**Comment by cgcgcg** [comment]
The more I think about it, the less I am convinced that this is correct. It will not bug out, as it did before, but it probably is still not quite correct. The graph that is used for the coloring is given in terms of local row and column IDs as well, so the coloring might be off. I'll look into it when I have time.

**Comment by srajama1** [comment]
The coloring is on symmetric graph and only on the local part. Assuming there is no unsymmetric ordering of local rows and local columns it should be fine. If row i and column i are not the same vertex then we are coloring what is just in the row adjacency and ignoring any mismatch in column adjacency.

**Comment by jhux2** [comment]
@srajama1 @cgcgcg How much work is involved in getting this merged?  Having MT Gauss-Seidel is important for the EMPIRE L2 milestone.

**Comment by cgcgcg** [comment]
@srajama1 Do you think this can be merged?

**Comment by srajama1** [comment]
Let us fix column map != row map outside KK. We can merge this.

**Pull request merged**